tailcfg/proto_port_range.go

```diff
@@ -0,0 +1,191 @@
+	Proto int
+	Ports PortRange
+}
+
+func (ppr ProtoPortRange) String() string {
+	if ppr.Proto == 0 {
+		if ppr.Ports == PortRangeAny {
+			return "*"
+		}
+	}
+	var buf strings.Builder
+	switch p := ipproto.Proto(ppr.Proto); p {
+	case
+		ipproto.ICMPv4,
+		ipproto.IGMP,
+		ipproto.ICMPv6,
+		ipproto.TCP,
+		ipproto.UDP,
+		ipproto.DCCP,
+		ipproto.GRE,
+		ipproto.SCTP:
+		fmt.Fprintf(&buf, "%s:", strings.ToLower(p.String()))
+
+	default:
+		if ppr.Proto != 0 {
+			fmt.Fprintf(&buf, "%d:", ppr.Proto)
+		}
+	}
+	pr := ppr.Ports
+	if pr.First == pr.Last {
+		fmt.Fprintf(&buf, "%d", pr.First)
+	} else if pr == PortRangeAny {
+		buf.WriteByte('*')
+	} else {
+		fmt.Fprintf(&buf, "%d-%d", pr.First, pr.Last)
+	}
+	return buf.String()
+}
+
+// ParseProtoPortRanges parses a slice of IP port range fields.
+func ParseProtoPortRanges(ips []string) ([]ProtoPortRange, error) {
+	var out []ProtoPortRange
+	for _, p := range ips {
+		ppr, err := parseProtoPortRange(p)
+		if err != nil {
+			return nil, err
+		}
+		out = append(out, *ppr)
+	}
+	return out, nil
+}
+
+func parseProtoPortRange(ipProtoPort string) (*ProtoPortRange, error) {
+	if ipProtoPort == "" {
+		return nil, errors.New("empty string")
+	}
+	if ipProtoPort == "*" {
+		return &ProtoPortRange{Ports: PortRangeAny}, nil
+	}
+	if !strings.Contains(ipProtoPort, ":") {
+		ipProtoPort = "*:" + ipProtoPort
+	}
+	protoStr, ports, err := parseHostPortRange(ipProtoPort)
+	if err != nil {
+		return nil, err
+	}
+	if protoStr == "" {
+		return nil, errors.New("empty protocol")
+	}
+	if len(ports) == 0 {
+		return nil, errors.New("empty port range")
+	}
+	if len(ports) > 1 {
+		return nil, errors.New("only one port range allowed")
+	}
+
+	ppr := &ProtoPortRange{
+		Ports: ports[0],
+	}
+	if protoStr == "*" {
+		return ppr, nil
+	}
+	ipProto, _, err := ipproto.ResolveProtoName(protoStr)
+	if err != nil {
+		return nil, err
+	}
+
+	ppr.Proto = int(ipProto)
+	return ppr, nil
+}
+
+// parseHostPortRange parses hostport as HOST:PORTS where HOST is
+// returned unchanged and PORTS is is either "*" or a comma-separated
+// list of PORTNUM or PORTLOW-PORTHIGH ranges.
+func parseHostPortRange(hostport string) (host string, ports []PortRange, err error) {
+	hostport = strings.ToLower(hostport)
+
+	// Need to find the *last* colon, because hostnames can contain
+	// user, group and tag names. For example:
+	//    tag:abc:1-3,10
+	// means
+	//    host: tag:abc
+	//    ports: 1, 2, 3, 10
+	colon := strings.LastIndexByte(hostport, ':')
+	if colon < 0 {
+		return "", nil, vizerror.New("hostport must contain a colon (\":\")")
+	}
+	host = hostport[:colon]
+	portlist := hostport[colon+1:]
+
+	if strings.Contains(host, ",") {
+		return "", nil, vizerror.New("host cannot contain a comma (\",\")")
+	}
+
+	if portlist == "*" {
+		// Special case: permit hostname:* as a port wildcard.
+		ports = append(ports, PortRangeAny)
+		return host, ports, nil
+	}
+
+	pl := strings.Split(portlist, ",")
```

Comment:
This was only relevant for parsing comma separated lists of ports like `80-90,443-450,600`. I don't think we need that here

Reply:
I would like to not duplicate this code, and this is a requirement of other locations in which it is used.